Dear Editor,

I am writing to submit our article entitled "A simple test to uncover signals of CpG hypermutability using posterior predictive simulations" for publication in your journal.

In our research group, we routinely use posterior predictive tests, which are a central tool in the Bayesian toolbox. These tests allow us to validate our phylogenetic substitution models, and to detect the presence of molecular evolutionary processes not accounted for by the substitution models, and thus potential confonders.

In this Research Note, we show that it is possible to detect the presence of CpG hypermutability with a simple test based on posterior predictive sampling using one of the most widely used phylogenetic models, namely the GTR+G substitution model. On the other hand, the interest of this test also stems from the fact that, knowing that CpG hypermutability is a confounding factor in the detection of negative and positive selection on codon usage as well as on amino acids, which is a key step in deciphering genomic information, this test should accompany any large-scale screening aimed at detecting negative or positive selection while there is a risk of methylation of cytosines within CpG contexts. In addition, the test could easily be extended to other contexts, such as the controversial one of TpA. BMC Research Notes seems well suited to publish our short note on this simple but valuable test.

Thank you for considering our submission.

Sincerely,
Simon Laurin-Lemay
